CSS counter-* should not eagerly evaluate calc()

weinig opened the pull request:
#### 5b689e6770dcfa551c1f493bcf9d1bee7c79c95c
<pre>
CSS counter-* should not eagerly evaluate calc()
<a href="https://bugs.webkit.org/show_bug.cgi?id=279326">https://bugs.webkit.org/show_bug.cgi?id=279326</a>

Reviewed by Darin Adler.

Replaces eager evaluation of calc() in counter-set, counter-increment, and
counter-reset, with storing the primitive value for resolution at style
building time.

Also took opportunity to move CSS List and CSS Counter Style consumers
to their own files.

* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-increment-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-increment-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-reset-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-reset-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-set-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-set-valid.html:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSCounterStyle.cpp:
* Source/WebCore/css/CSSCounterStyleRule.cpp:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Lists.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Lists.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/process-css-properties.py:

Canonical link: <a href="https://commits.webkit.org/283340@main">https://commits.webkit.org/283340@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/682423a1a254c02e3b919963938f50d13582d171

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66003 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45376 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18622 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70033 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16612 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68121 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53175 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16894 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52976 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11560 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69070 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41871 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57142 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33611 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38542 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14519 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15488 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60421 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14866 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71736 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9958 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14279 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60292 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9990 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57204 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60582 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14558 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8225 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1866 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41184 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42260 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43443 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42004 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.cpp:
```diff
@@ -0,0 +1,298 @@
+        if (!symbol)
+            return nullptr;
+
+        // Additive tuples must be specified in order of strictly descending weight.
```
Do we need test cases for how this works with `calc`? Seems subtle.

weinig replied:
I am not yet sure how this is supposed to work. Since this is an @counter-style rule descriptor, the usual time for calc() resolution does not apply and https://drafts.csswg.org/css-counter-styles-3/ doesn't define what to do.

I have filed https://github.com/w3c/csswg-drafts/issues/10853 to start the conversation with editors about this class of problem.

For now, this is just moved code, so nothing is needed in this PR.

weinig replied:
(adding those "Deprecated" named functions was to make it really easy to find all the places, like this, that we need to figure out).

darinadler reviewed (approved):
I didn't check why the compile is failing but I approve once you resolve that and get the tests to pass.

weinig committed:
CSS counter-* should not eagerly evaluate calc()
https://bugs.webkit.org/show_bug.cgi?id=279326

Reviewed by Darin Adler.

Replaces eager evaluation of calc() in counter-set, counter-increment, and
counter-reset, with storing the primitive value for resolution at style
building time.

Also took opportunity to move CSS List and CSS Counter Style consumers
to their own files.

* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-increment-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-increment-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-reset-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-reset-valid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-set-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-lists/parsing/counter-set-valid.html:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSCounterStyle.cpp:
* Source/WebCore/css/CSSCounterStyleRule.cpp:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Lists.cpp: Added.
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Lists.h: Added.
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/css/process-css-properties.py:

Canonical link: https://commits.webkit.org/283340@main
